### ISSUE TITLE
Inconsistent Source Map generation based on file path depth

### ISSUE THREAD

**Issue by npetruzzelli** [floridoo/vinyl-sourcemaps-apply#12]
Hi! I'm new to GitHub and I look forward to working with you!

Quite a few details already exist in pull request: https://github.com/sindresorhus/gulp-autoprefixer/pull/66

Right now I can't pin point a cause between `PostCSS`, `gulp-autoprefixer`, and `vinyl-sourcemaps-apply`. @sindresorhus suggested I move the conversation here. 

I've just started looking at the source code for this repository, I'll hopefully have more to say than "my stuff is broke!" with in a couple days. I wanted to get the conversation started, since during the week we are all extra busy. This way it may at least be sitting in the back of your mind.

I'll re-post some of the details here to keep everything in one place.

**The problem doesn't happen for files at the top level.**

> Consider the following is the project folder structure.
> 
> ```
> source/
>     sub-directory/
>         app.scss
>     top.scss
> gulpfile.js
> package.json
> ```
> 
> `npm i --save-dev gulp gulp-autoprefixer gulp-load-plugins gulp-plumber gulp-sass gulp-sourcemaps`
> 
> ``` json
> {
>   "private":true,
>   "name": "test-sourcemaps",
>   "version": "1.0.0",
>   "devDependencies": {
>     "gulp": "^3.9.1",
>     "gulp-autoprefixer": "^3.1.0",
>     "gulp-load-plugins": "^1.2.4",
>     "gulp-plumber": "^1.1.0",
>     "gulp-sass": "^2.3.2",
>     "gulp-sourcemaps": "^1.6.0"
>   }
> }
> ```
> 
> ``` javascript
> /* ./gulpfile.js */
> const gulp            = require('gulp')
>     , gulpLoadPlugins = require('gulp-load-plugins')
>     , $               = gulpLoadPlugins()
> ;
> 
> gulp.task('default', (callback)=>{
>     return gulp.src('source/**/*.scss', {
>             base: 'source'
>         })
>         .pipe($.sourcemaps.init())
>         .pipe($.sass.sync({
>             outputStyle: 'expanded',
>             precision: 10,
>             includePaths: ['.']
>         }).on('error', $.sass.logError))
>         .pipe($.autoprefixer())
>         .pipe($.sourcemaps.write('.', {
>             includeContent: false
>         }))
>         .pipe(gulp.dest('dist/'))
>     ;
> });
> ```
> 
> ``` sass
> /* ./source/top.scss */
> $bgColor: cyan;
> 
> body{
>     background:$bgColor;
> }
> ```
> 
> ``` sass
> /* ./source/sub-directory/app.scss */
> $bgColor: lime;
> 
> body{
>     background:$bgColor;
> }
> ```
> 
> Modify `gulp-autoprefixer` to track it: https://github.com/sindresorhus/gulp-autoprefixer/blob/v3.1.0/index.js
> 
> ``` diff
>           if (file.isStream()) {
>               cb(new gutil.PluginError('gulp-autoprefixer', 'Streaming not supported'));
>               return;
>           }
> 
>           postcss(autoprefixer(opts)).process(file.contents.toString(), {
>               map: file.sourceMap ? {annotation: false} : false,
>               from: file.path,
>               to: file.path
>           }).then(function (res) {
>               file.contents = new Buffer(res.css);
> 
> ++            console.log('GULP-AUTOPREFIXER #1');
> ++            console.log(file.sourceMap);
>               if (res.map && file.sourceMap) {
>                   applySourceMap(file, res.map.toString());
>               }
> ++            console.log('GULP-AUTOPREFIXER #2');
> ++            console.log(file.sourceMap);
> ```
> 
> The console will display the following: (using diff for convenient highlighting.)
> 
> ``` diff
>   GULP-AUTOPREFIXER #1
>   { version: 3,
>     file: 'top.css',
>     sources: [ 'top.scss' ],
>     sourcesContent: [ '$bgColor : cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
>     mappings: 'AAEA,AAAA,IAAI,CAAA;EACA,UAAU,EAHH,IAAI;CAId',
>     names: [] }
>   GULP-AUTOPREFIXER #2
>   { version: 3,
>     sources: [ 'top.scss' ],
>     names: [],
>     mappings: 'AAEA;EACI,iBAHW;CAId',
>     file: 'top.css',
>     sourcesContent: [ '$bgColor : cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }
>   GULP-AUTOPREFIXER #1
>   { version: 3,
>     file: 'sub-directory/app.css',
> ++  sources: [ 'sub-directory/app.scss' ],
> ++  sourcesContent: [ '$bgColor : lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
>     mappings: 'AAEA,AAAA,IAAI,CAAA;EACA,UAAU,EAHH,IAAI;CAId',
>     names: [] }
>   GULP-AUTOPREFIXER #2
>   { version: 3,
> --  sources: [ 'app.css' ],
>     names: [],
>     mappings: 'AAAA;EACE,iBAAiB;CAClB',
>     file: 'app.css',
> --  sourcesContent: [ 'body {\n  background: lime;\n}\n' ] }
> 
> ```
> 
> We get the following:
> 
> **top.css.map**
> 
> ``` json
> {"version":3,"sources":["top.scss"],"names":[],"mappings":"AAEA;EACI,iBAHW;CAId","file":"top.css"}
> ```
> 
> **app.css.map**
> 
> ``` json
> {"version":3,"sources":["app.css"],"names":[],"mappings":"AAAA;EACE,iBAAiB;CAClB","file":"sub-directory/app.css"}
> ```
> 
> `top.css.map` still points to the SASS file, while `app.css.map` thinks that the output from `Autoprefixer` is the source.

**Comment by npetruzzelli** [comment]
Hi, its been about two weeks. I'm still looking at the code to see what might be going on. I haven't been able to put as much time into as I would have liked.

My initial comparison between the two is as follows:
-   **Variable type**
  -   [`vinyl-sourcemaps-apply` (VSA) works with JSON objects](https://github.com/floridoo/vinyl-sourcemaps-apply/blob/v0.2.1/index.js#L6-L8).
  -   [`PostCSS` works with JSON strings](https://github.com/postcss/postcss/blob/master/lib/previous-map.es6#L98-L119). I don't know if it is project preference or if there is a specific driver for this, but [this may have to do with this](https://github.com/postcss/postcss/blob/5.1.2/lib/map-generator.es6#L28-L30).
-   **`SourceMapGenerator`**
  -   [VSA creates its generator](https://github.com/floridoo/vinyl-sourcemaps-apply/blob/v0.2.1/index.js#L26) using the [`SourceMapGenerator.fromSourceMap` method](https://github.com/mozilla/source-map/tree/0.5.6#sourcemapgeneratorfromsourcemapsourcemapconsumer),
  -   [`PostCSS` passes an object](https://github.com/postcss/postcss/blob/5.1.2/lib/map-generator.es6#L194) to a [`new SourceMapGenerator`](https://github.com/mozilla/source-map/tree/0.5.6#new-sourcemapgeneratorstartofsourcemap). The object contains just a [file system path](https://github.com/postcss/postcss/blob/5.1.2/lib/map-generator.es6#L144-L151).
-   **`applySourceMap`**
  -   [VSA](https://github.com/floridoo/vinyl-sourcemaps-apply/blob/v0.2.1/index.js#L25-L28) - consumes a SourceMapConsumer
  -   [`PostCSS`](https://github.com/postcss/postcss/blob/5.1.2/lib/map-generator.es6#L93-L110) - consumes a SourceMapConsumer, indicates the source file name, and indicates the source file map path.
    -   If your wondering about `prev.consumer()`, see: https://github.com/postcss/postcss/blob/5.1.2/lib/previous-map.es6#L35-L49

The following is a rough list of what imports/requires what in `PostCSS`, leading up from Mozilla's `source-map` module. Modules imported by more than one other module have those modules listed in parenthesis. All modules refer to the `.es6` files in the `lib` directory of the `PostCSS` repository.

```
postcss
    root (postcss, parser)
        lazy-result (root, processor)
            parse (postcss, lazy-result)
                input
                    previous-map
                        source-map
            map-generator
                source-map

parser (parse)
processor (postcss, root)
```

I apologize if the comparison against PostCSS seems a bit excessive, but I am trying to understand each step of the process so that I can help contribute to the solution.

Next, I plan to debug the value of the source map as it passes through VSA and see where it starts producing the unexpected result. I'll probably end up digging into the code for [`source-map`](https://github.com/mozilla/source-map/) as well to better understand it all.

Have you been able to have a look into it or have any suggestions to narrow things down, @floridoo ? Have I missed reporting information required when opening an issue against this repository?

**Comment by npetruzzelli** [comment]
# TL;DR

Switching from `gulp-autoprefixer` to `gulp-postcss` + `autoprefixer` produces functional source maps, though I have my concerns about the `file` property of the resulting maps.

It doesn't "fix" or even identify the source of the original problem, but it does work.

**Comment by npetruzzelli** [comment]
# Details

I've done some more digging, some head scratching, and then some digging over something I could swear I had already looked at.

**node_modules/gulp-autoprefixer/index.js**

``` diff
              console.log('GULP-AUTOPREFIXER #1');
              console.log(file.sourceMap);
++            console.log('GULP-AUTOPREFIXER RES.MAP', JSON.parse(res.map.toString()) );
              if (res.map && file.sourceMap) {
                  applySourceMap(file, res.map.toString());
              }
              console.log('GULP-AUTOPREFIXER #2');
              console.log(file.sourceMap);
++            console.log('\nGULP-AUTOPREFIXER DONE\n');
```

``` diff
--/* ./source/top.scss */
++/* ./source/top.scss               */
  $bgColor: cyan;

  body{
      background:$bgColor;
  }
```

Why the extra spaces in the first line of `top.scss`? To make it identical (in column count) to `app.scss`.

The resulting console output is: 

``` diff
  GULP-AUTOPREFIXER #1
  { version: 3,
    file: 'top.css',
    sources: [ 'top.scss' ],
    sourcesContent: [ '/* ./source/top.scss               */\r\n$bgColor: cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
    mappings: 'AAAA,qCAAqC;AAGrC,AAAA,IAAI,CAAA;EACA,UAAU,EAHJ,IAAI;CAIb',
    names: [] }
  GULP-AUTOPREFIXER RES.MAP { version: 3,
    sources: [ 'top.css' ],
    names: [],
    mappings: 'AAAA,qCAAqC;AACrC;EACE,iBAAiB;CAClB',
    file: 'top.css',
    sourcesContent: [ '/* ./source/top.scss               */\nbody {\n  background: cyan;\n}\n' ] }
  GULP-AUTOPREFIXER #2
  { version: 3,
    sources: [ 'top.scss' ],
    names: [],
    mappings: 'AAAA,qCAAqC;AAGrC;EACI,iBAHU;CAIb',
    file: 'top.css',
    sourcesContent: [ '/* ./source/top.scss               */\r\n$bgColor: cyan;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ]   }

  GULP-AUTOPREFIXER DONE

  GULP-AUTOPREFIXER #1
  { version: 3,
    file: 'sub-directory/app.css',
    sources: [ 'sub-directory/app.scss' ],
    sourcesContent: [ '/* ./source/sub-directory/app.scss */\r\n$bgColor: lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ],
    mappings: 'AAAA,qCAAqC;AAGrC,AAAA,IAAI,CAAA;EACA,UAAU,EAHJ,IAAI;CAIb',
    names: [] }
  GULP-AUTOPREFIXER RES.MAP { version: 3,
    sources: [ 'app.css' ],
    names: [],
    mappings: 'AAAA,qCAAqC;AACrC;EACE,iBAAiB;CAClB',
    file: 'app.css',
    sourcesContent: [ '/* ./source/sub-directory/app.scss */\nbody {\n  background: lime;\n}\n' ] }
  GULP-AUTOPREFIXER #2
  { version: 3,
--  sources: [ 'app.css' ],
    names: [],
--  mappings: 'AAAA,qCAAqC;AACrC;EACE,iBAAiB;CAClB',
    file: 'app.css',
--  sourcesContent: [ '/* ./source/sub-directory/app.scss */\nbody {\n  background: lime;\n}\n' ] }

  GULP-AUTOPREFIXER DONE
```

The expected/desired output would be:

``` diff
  { version: 3,
++  sources: [ 'sub-directory/app.scss' ],
    names: [],
++  mappings: 'AAAA,qCAAqC;AAGrC;EACI,iBAHU;CAIb',
    file: 'app.css',
++  sourcesContent: [ '/* ./source/sub-directory/app.scss */\r\n$bgColor: lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }

```

You can see that `sources` and `sourcesContent` would be correct and that `mappings` would both be correct and be identical, as expected, to `top.css.map`.

As far as I can tell, both `PostCSS` and `vinyl-sourcemaps-apply` are providing the arguments that are expected by `source-map`, even if both are consuming it in different ways.

When using the modification to `gulp-autoprefixer` that I suggested in the pull request for `gulp-autoprefixer`, the console output becomes:

```
GULP-AUTOPREFIXER RES.MAP { version: 3,
  sources: [ 'sub-directory/app.scss' ],
  names: [],
  mappings: 'AAAA,qCAAqC;AAGrC;EACI,iBAHU;CAIb',
  file: 'app.css',
  sourcesContent: [ '/* ./source/sub-directory/app.scss */\r\n$bgColor: lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }
GULP-AUTOPREFIXER #2
{ version: 3,
  sources: [ 'sub-directory/app.scss' ],
  names: [],
  mappings: 'AAAA,qCAAqC;AAGrC;EACI,iBAHU;CAIb',
  file: 'app.css',
  sourcesContent: [ '/* ./source/sub-directory/app.scss */\r\n$bgColor: lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }
```

In the end, I don't know enough to say where the error is happening and diving through the source code of five different repositories (`gulp-autoprefixer`, `autoprefixer`, `PostCSS`, `vinyl-sourcemaps-apply`, and `source-map`) is a time devouring task. In the process, I learned that PostCSS has a gulp-plugin of its own. After testing it with a similar setup, it produced a working source map.

**package.json**

``` diff
  {
    "private": true,
    "name": "test-sourcemaps",
    "version": "1.0.0",
    "devDependencies": {
++    "autoprefixer": "^6.4.0",
++    "browser-sync": "^2.14.0",
      "gulp": "^3.9.1",
--    "gulp-autoprefixer": "^3.1.0",
      "gulp-load-plugins": "^1.2.4",
      "gulp-plumber": "^1.1.0",
++    "gulp-postcss": "^6.1.1",
      "gulp-sass": "^2.3.2",
      "gulp-sourcemaps": "^1.6.0"
    }
  }
```

**gulpfile.js**

``` diff
  /* ./gulpfile.js */
  const gulp            = require('gulp')
      , gulpLoadPlugins = require('gulp-load-plugins')
++    , autoprefixer    = require('autoprefixer')
++    , browserSync     = require('browser-sync')
      , $               = gulpLoadPlugins()
  ;

  gulp.task('default', (callback)=>{
++    var postCssProcessors = [
++            autoprefixer()
++        ]
++        , postCssOptions = {}
++    ;
      return gulp.src('source/**/*.scss', {
              base: 'source'
          })
          .pipe($.sourcemaps.init())
          .pipe($.sass.sync({
              outputStyle: 'expanded',
              precision: 10,
              includePaths: ['.']
          }).on('error', $.sass.logError))
--        .pipe($.autoprefixer())
++        .pipe($.postcss(postCssProcessors, postCssOptions))
          .pipe($.sourcemaps.write('.', {
              includeContent: false
++            , sourceRoot: '/source/'
          }))
          .pipe(gulp.dest('dist/'))
      ;
  });

++gulp.task('serve', () => {
++  browserSync({
++    notify: false,
++    port: 9000,
++    server: {
++      baseDir: ['dist'],
++      routes: {
++        '/source': 'source'
++      }
++    }
++  });
++});
```

The resulting map is:

```
{ version: 3,
  sources: [ 'sub-directory/app.scss' ],
  names: [],
  mappings: 'AAAA,qCAAqC;AAGrC;EACI,iBAHU;CAIb',
  file: 'sub-directory/app.css',
  sourcesContent: [ '/* ./source/sub-directory/app.scss */\r\n$bgColor: lime;\r\n\r\nbody{\r\n    background:$bgColor;\r\n}' ] }
```

I thought that the `file` property was supposed to be a file name, not a file path, or at least a path relative to the map file itself, not [relative](https://github.com/postcss/gulp-postcss/blob/6.1.1/index.js#L60) to the file [base](https://github.com/gulpjs/vinyl/blob/v1.2.0/index.js#L186). The [Source Maps Revision 3 Proposal](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/) does give guidance on `sources`, but not `file`. I was expecting `file: 'app.css'`. The proposal describes `file` as _"An optional name of the generated code that this source map is associated with."_ Otherwise, the source map appears correct and, more importantly, functions correctly in the browser.

I'm confused that sources are relative to the `source` directory, even when you remove `base: 'source'`. Isn't the default base directory the current working directory? Aren't sources supposed to be relative to the map file? This isn't a big deal, since this can probably be dealt with using `destPath` and/or `mapSources` in the write options for `gulp-sourcemaps`. The current sources can still work with a root relative URL (["path-absolute URL"](https://url.spec.whatwg.org/#syntax-url-path-absolute)) for `sourceRoot`.

I've asked a lot of questions that begin to stray from the original post. Should I have the time to revisit this, I'll find the appropriate place to do so. It doesn't feel bound to this repository.

I'm closing the issue since I now have a working build and it isn't clear whether or not `vinyl-sourcemaps-apply` was the root of the problem to begin with.
